packages/visual-sdk/src/store/EditorEntityStore.ts

```diff
@@ -0,0 +1,169 @@
+import type { Asset, Entry } from 'contentful';
+
+import { EntityStore } from './EntityStore';
+
+export type RequestEntitiesMessage = {
+  entityIds: string[];
+  entityType: 'Asset' | 'Entry';
+  locale: string;
+};
+
+export type RequestedEntitiesMessage = {
+  entities: Array<Entry | Asset>;
+};
+
+export enum PostMessageMethods {
+  REQUEST_ENTITIES = 'REQUEST_ENTITIES',
+  REQUESTED_ENTITIES = 'REQUESTED_ENTITIES',
+}
+
+type SendMessage = (
+  method: PostMessageMethods.REQUEST_ENTITIES,
+  params: RequestEntitiesMessage
+) => void;
+type Subscribe = (
+  method: PostMessageMethods.REQUESTED_ENTITIES,
+  cb: (message: RequestedEntitiesMessage) => void
+) => VoidFunction;
+
+/**
+ * EntityStore which resolves entries and assets from the editor
+ * over the sendMessage and subscribe functions.
+ */
+export class EditorEntityStore extends EntityStore {
+  private requestCache = new Map<string, Promise<any>>();
+  private sendMessage: SendMessage;
+  private subscribe: Subscribe;
+  private timeoutDuration: number;
+
+  constructor({
+    entities,
+    locale,
+    sendMessage,
+    subscribe,
+    timeoutDuration = 3000,
+  }: {
+    entities: Array<Entry | Asset>;
+    locale: string;
+    sendMessage: SendMessage;
+    subscribe: Subscribe;
+    timeoutDuration?: number;
+  }) {
+    super({ entities, locale });
+    this.sendMessage = sendMessage;
+    this.subscribe = subscribe;
+    this.timeoutDuration = timeoutDuration;
+  }
+
+  private cleanupPromise(referenceId: string) {
+    setTimeout(() => {
+      this.requestCache.delete(referenceId);
+    }, 300);
+  }
+
+  private cacheIdSeperator = ',';
+
+  private getCacheId(id: string[]): string {
+    return id.length === 1 ? id[0] : id.join(this.cacheIdSeperator);
+  }
+
+  private findMissingEntites(ids: string[]) {
+    const missing = [];
+
+    for (const id of ids) {
+      const entry = this.entitiesMap.get(id);
+      if (!entry) {
+        missing.push(id);
+      }
+    }
+
+    return missing;
+  }
+
+  private async fetchEntity(ids: string[]): Promise<Array<Entry>>;
```

Comment:
I find this code a bit confusing, could explain what's happening here? Why do we have 3 times `fetchEntity`? Can we not have one generic function and then fetchAsset?

Reply:
Ah that's for typing ([function overload](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads))- so you get the correct type

if you call it without  `isAsset` you get the `Entry` type
if you call it with `isAsset` you get the `Asset` type
and the third one is then the implementation, which needs to implement both overloads